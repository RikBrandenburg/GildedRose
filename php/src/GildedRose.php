<?php

declare(strict_types=1);

namespace GildedRose;

final class GildedRose
{
    /**
     * @param Item[] $items
     */
    public function __construct(
        private array $items
    ) {
    }

    /**
     * Update the quality of all items in the Gilded Rose inventory.
     */
    public function updateQuality(): void
    {
        foreach ($this->items as $item) {
            $wrapper = ItemFactory::create($item);
            $wrapper->update();
        }
    }
}
