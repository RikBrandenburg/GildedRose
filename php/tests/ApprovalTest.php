<?php

declare(strict_types=1);

namespace Tests;

use ApprovalTests\Approvals;
use GildedRose\GildedRose;
use GildedRose\Item;
use PHPUnit\Framework\TestCase;

/**
 * This unit test uses [Approvals](https://github.com/approvals/ApprovalTests.php).
 *
 * There are two test cases here with different styles:
 * <li>"foo" is more similar to the unit test from the 'Java' version
 * <li>"thirtyDays" is more similar to the TextTest from the 'Java' version
 *
 * I suggest choosing one style to develop and deleting the other.
 */
class ApprovalTest extends TestCase
{
    public function testFoo(): void
    {
        $items = [new Item('foo', 0, 0)];
        $app = new GildedRose($items);
        $app->updateQuality();

        $itemStrings = array_map(
            fn($item) => (string) $item,
            $items
        );

        Approvals::verifyList($itemStrings);
    }

    public function testThirtyDays(): void
    {
        $items = [
            new Item("+5 Dexterity Vest", 10, 20),
            new Item("Aged Brie", 2, 0),
            new Item("Elixir of the Mongoose", 5, 7),
            new Item("Sulfuras, Hand of Ragnaros", 0, 80),
            new Item("Sulfuras, Hand of Ragnaros", -1, 80),
            new Item("Backstage passes to a TAFKAL80ETC concert", 15, 20),
            new Item("Backstage passes to a TAFKAL80ETC concert", 10, 49),
            new Item("Backstage passes to a TAFKAL80ETC concert", 5, 49),
            new Item("Conjured Mana Cake", 3, 6),
        ];

        $app = new GildedRose($items);

        $results = [];

        for ($day = 0; $day < 30; $day++) {
            $results[] = "-------- day $day --------";
            $results[] = "name, sellIn, quality";

            foreach ($items as $item) {
                $results[] = (string) $item;
            }

            $results[] = ""; // empty line between days

            $app->updateQuality();
        }

        Approvals::verifyList($results);
    }
}
