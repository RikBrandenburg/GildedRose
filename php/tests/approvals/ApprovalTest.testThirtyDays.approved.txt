[0] -> -------- day 0 --------
[1] -> name, sellIn, quality
[2] -> +5 Dexterity Vest, 10, 20
[3] -> Aged Brie, 2, 0
[4] -> Elixir of the Mongoose, 5, 7
[5] -> Sulfuras, Hand of Ragnaros, 0, 80
[6] -> Sulfuras, Hand of Ragnaros, -1, 80
[7] -> Backstage passes to a TAFKAL80ETC concert, 15, 20
[8] -> Backstage passes to a TAFKAL80ETC concert, 10, 49
[9] -> Backstage passes to a TAFKAL80ETC concert, 5, 49
[10] -> Conjured Mana Cake, 3, 6
[11] -> 
[12] -> -------- day 1 --------
[13] -> name, sellIn, quality
[14] -> +5 Dexterity Vest, 9, 19
[15] -> Aged Brie, 1, 1
[16] -> Elixir of the Mongoose, 4, 6
[17] -> Sulfuras, Hand of Ragnaros, 0, 80
[18] -> Sulfuras, Hand of Ragnaros, -1, 80
[19] -> Backstage passes to a TAFKAL80ETC concert, 14, 21
[20] -> Backstage passes to a TAFKAL80ETC concert, 9, 50
[21] -> Backstage passes to a TAFKAL80ETC concert, 4, 50
[22] -> Conjured Mana Cake, 2, 4
[23] -> 
[24] -> -------- day 2 --------
[25] -> name, sellIn, quality
[26] -> +5 Dexterity Vest, 8, 18
[27] -> Aged Brie, 0, 2
[28] -> Elixir of the Mongoose, 3, 5
[29] -> Sulfuras, Hand of Ragnaros, 0, 80
[30] -> Sulfuras, Hand of Ragnaros, -1, 80
[31] -> Backstage passes to a TAFKAL80ETC concert, 13, 22
[32] -> Backstage passes to a TAFKAL80ETC concert, 8, 50
[33] -> Backstage passes to a TAFKAL80ETC concert, 3, 50
[34] -> Conjured Mana Cake, 1, 2
[35] -> 
[36] -> -------- day 3 --------
[37] -> name, sellIn, quality
[38] -> +5 Dexterity Vest, 7, 17
[39] -> Aged Brie, -1, 4
[40] -> Elixir of the Mongoose, 2, 4
[41] -> Sulfuras, Hand of Ragnaros, 0, 80
[42] -> Sulfuras, Hand of Ragnaros, -1, 80
[43] -> Backstage passes to a TAFKAL80ETC concert, 12, 23
[44] -> Backstage passes to a TAFKAL80ETC concert, 7, 50
[45] -> Backstage passes to a TAFKAL80ETC concert, 2, 50
[46] -> Conjured Mana Cake, 0, 0
[47] -> 
[48] -> -------- day 4 --------
[49] -> name, sellIn, quality
[50] -> +5 Dexterity Vest, 6, 16
[51] -> Aged Brie, -2, 6
[52] -> Elixir of the Mongoose, 1, 3
[53] -> Sulfuras, Hand of Ragnaros, 0, 80
[54] -> Sulfuras, Hand of Ragnaros, -1, 80
[55] -> Backstage passes to a TAFKAL80ETC concert, 11, 24
[56] -> Backstage passes to a TAFKAL80ETC concert, 6, 50
[57] -> Backstage passes to a TAFKAL80ETC concert, 1, 50
[58] -> Conjured Mana Cake, -1, 0
[59] -> 
[60] -> -------- day 5 --------
[61] -> name, sellIn, quality
[62] -> +5 Dexterity Vest, 5, 15
[63] -> Aged Brie, -3, 8
[64] -> Elixir of the Mongoose, 0, 2
[65] -> Sulfuras, Hand of Ragnaros, 0, 80
[66] -> Sulfuras, Hand of Ragnaros, -1, 80
[67] -> Backstage passes to a TAFKAL80ETC concert, 10, 25
[68] -> Backstage passes to a TAFKAL80ETC concert, 5, 50
[69] -> Backstage passes to a TAFKAL80ETC concert, 0, 50
[70] -> Conjured Mana Cake, -2, 0
[71] -> 
[72] -> -------- day 6 --------
[73] -> name, sellIn, quality
[74] -> +5 Dexterity Vest, 4, 14
[75] -> Aged Brie, -4, 10
[76] -> Elixir of the Mongoose, -1, 0
[77] -> Sulfuras, Hand of Ragnaros, 0, 80
[78] -> Sulfuras, Hand of Ragnaros, -1, 80
[79] -> Backstage passes to a TAFKAL80ETC concert, 9, 27
[80] -> Backstage passes to a TAFKAL80ETC concert, 4, 50
[81] -> Backstage passes to a TAFKAL80ETC concert, -1, 0
[82] -> Conjured Mana Cake, -3, 0
[83] -> 
[84] -> -------- day 7 --------
[85] -> name, sellIn, quality
[86] -> +5 Dexterity Vest, 3, 13
[87] -> Aged Brie, -5, 12
[88] -> Elixir of the Mongoose, -2, 0
[89] -> Sulfuras, Hand of Ragnaros, 0, 80
[90] -> Sulfuras, Hand of Ragnaros, -1, 80
[91] -> Backstage passes to a TAFKAL80ETC concert, 8, 29
[92] -> Backstage passes to a TAFKAL80ETC concert, 3, 50
[93] -> Backstage passes to a TAFKAL80ETC concert, -2, 0
[94] -> Conjured Mana Cake, -4, 0
[95] -> 
[96] -> -------- day 8 --------
[97] -> name, sellIn, quality
[98] -> +5 Dexterity Vest, 2, 12
[99] -> Aged Brie, -6, 14
[100] -> Elixir of the Mongoose, -3, 0
[101] -> Sulfuras, Hand of Ragnaros, 0, 80
[102] -> Sulfuras, Hand of Ragnaros, -1, 80
[103] -> Backstage passes to a TAFKAL80ETC concert, 7, 31
[104] -> Backstage passes to a TAFKAL80ETC concert, 2, 50
[105] -> Backstage passes to a TAFKAL80ETC concert, -3, 0
[106] -> Conjured Mana Cake, -5, 0
[107] -> 
[108] -> -------- day 9 --------
[109] -> name, sellIn, quality
[110] -> +5 Dexterity Vest, 1, 11
[111] -> Aged Brie, -7, 16
[112] -> Elixir of the Mongoose, -4, 0
[113] -> Sulfuras, Hand of Ragnaros, 0, 80
[114] -> Sulfuras, Hand of Ragnaros, -1, 80
[115] -> Backstage passes to a TAFKAL80ETC concert, 6, 33
[116] -> Backstage passes to a TAFKAL80ETC concert, 1, 50
[117] -> Backstage passes to a TAFKAL80ETC concert, -4, 0
[118] -> Conjured Mana Cake, -6, 0
[119] -> 
[120] -> -------- day 10 --------
[121] -> name, sellIn, quality
[122] -> +5 Dexterity Vest, 0, 10
[123] -> Aged Brie, -8, 18
[124] -> Elixir of the Mongoose, -5, 0
[125] -> Sulfuras, Hand of Ragnaros, 0, 80
[126] -> Sulfuras, Hand of Ragnaros, -1, 80
[127] -> Backstage passes to a TAFKAL80ETC concert, 5, 35
[128] -> Backstage passes to a TAFKAL80ETC concert, 0, 50
[129] -> Backstage passes to a TAFKAL80ETC concert, -5, 0
[130] -> Conjured Mana Cake, -7, 0
[131] -> 
[132] -> -------- day 11 --------
[133] -> name, sellIn, quality
[134] -> +5 Dexterity Vest, -1, 8
[135] -> Aged Brie, -9, 20
[136] -> Elixir of the Mongoose, -6, 0
[137] -> Sulfuras, Hand of Ragnaros, 0, 80
[138] -> Sulfuras, Hand of Ragnaros, -1, 80
[139] -> Backstage passes to a TAFKAL80ETC concert, 4, 38
[140] -> Backstage passes to a TAFKAL80ETC concert, -1, 0
[141] -> Backstage passes to a TAFKAL80ETC concert, -6, 0
[142] -> Conjured Mana Cake, -8, 0
[143] -> 
[144] -> -------- day 12 --------
[145] -> name, sellIn, quality
[146] -> +5 Dexterity Vest, -2, 6
[147] -> Aged Brie, -10, 22
[148] -> Elixir of the Mongoose, -7, 0
[149] -> Sulfuras, Hand of Ragnaros, 0, 80
[150] -> Sulfuras, Hand of Ragnaros, -1, 80
[151] -> Backstage passes to a TAFKAL80ETC concert, 3, 41
[152] -> Backstage passes to a TAFKAL80ETC concert, -2, 0
[153] -> Backstage passes to a TAFKAL80ETC concert, -7, 0
[154] -> Conjured Mana Cake, -9, 0
[155] -> 
[156] -> -------- day 13 --------
[157] -> name, sellIn, quality
[158] -> +5 Dexterity Vest, -3, 4
[159] -> Aged Brie, -11, 24
[160] -> Elixir of the Mongoose, -8, 0
[161] -> Sulfuras, Hand of Ragnaros, 0, 80
[162] -> Sulfuras, Hand of Ragnaros, -1, 80
[163] -> Backstage passes to a TAFKAL80ETC concert, 2, 44
[164] -> Backstage passes to a TAFKAL80ETC concert, -3, 0
[165] -> Backstage passes to a TAFKAL80ETC concert, -8, 0
[166] -> Conjured Mana Cake, -10, 0
[167] -> 
[168] -> -------- day 14 --------
[169] -> name, sellIn, quality
[170] -> +5 Dexterity Vest, -4, 2
[171] -> Aged Brie, -12, 26
[172] -> Elixir of the Mongoose, -9, 0
[173] -> Sulfuras, Hand of Ragnaros, 0, 80
[174] -> Sulfuras, Hand of Ragnaros, -1, 80
[175] -> Backstage passes to a TAFKAL80ETC concert, 1, 47
[176] -> Backstage passes to a TAFKAL80ETC concert, -4, 0
[177] -> Backstage passes to a TAFKAL80ETC concert, -9, 0
[178] -> Conjured Mana Cake, -11, 0
[179] -> 
[180] -> -------- day 15 --------
[181] -> name, sellIn, quality
[182] -> +5 Dexterity Vest, -5, 0
[183] -> Aged Brie, -13, 28
[184] -> Elixir of the Mongoose, -10, 0
[185] -> Sulfuras, Hand of Ragnaros, 0, 80
[186] -> Sulfuras, Hand of Ragnaros, -1, 80
[187] -> Backstage passes to a TAFKAL80ETC concert, 0, 50
[188] -> Backstage passes to a TAFKAL80ETC concert, -5, 0
[189] -> Backstage passes to a TAFKAL80ETC concert, -10, 0
[190] -> Conjured Mana Cake, -12, 0
[191] -> 
[192] -> -------- day 16 --------
[193] -> name, sellIn, quality
[194] -> +5 Dexterity Vest, -6, 0
[195] -> Aged Brie, -14, 30
[196] -> Elixir of the Mongoose, -11, 0
[197] -> Sulfuras, Hand of Ragnaros, 0, 80
[198] -> Sulfuras, Hand of Ragnaros, -1, 80
[199] -> Backstage passes to a TAFKAL80ETC concert, -1, 0
[200] -> Backstage passes to a TAFKAL80ETC concert, -6, 0
[201] -> Backstage passes to a TAFKAL80ETC concert, -11, 0
[202] -> Conjured Mana Cake, -13, 0
[203] -> 
[204] -> -------- day 17 --------
[205] -> name, sellIn, quality
[206] -> +5 Dexterity Vest, -7, 0
[207] -> Aged Brie, -15, 32
[208] -> Elixir of the Mongoose, -12, 0
[209] -> Sulfuras, Hand of Ragnaros, 0, 80
[210] -> Sulfuras, Hand of Ragnaros, -1, 80
[211] -> Backstage passes to a TAFKAL80ETC concert, -2, 0
[212] -> Backstage passes to a TAFKAL80ETC concert, -7, 0
[213] -> Backstage passes to a TAFKAL80ETC concert, -12, 0
[214] -> Conjured Mana Cake, -14, 0
[215] -> 
[216] -> -------- day 18 --------
[217] -> name, sellIn, quality
[218] -> +5 Dexterity Vest, -8, 0
[219] -> Aged Brie, -16, 34
[220] -> Elixir of the Mongoose, -13, 0
[221] -> Sulfuras, Hand of Ragnaros, 0, 80
[222] -> Sulfuras, Hand of Ragnaros, -1, 80
[223] -> Backstage passes to a TAFKAL80ETC concert, -3, 0
[224] -> Backstage passes to a TAFKAL80ETC concert, -8, 0
[225] -> Backstage passes to a TAFKAL80ETC concert, -13, 0
[226] -> Conjured Mana Cake, -15, 0
[227] -> 
[228] -> -------- day 19 --------
[229] -> name, sellIn, quality
[230] -> +5 Dexterity Vest, -9, 0
[231] -> Aged Brie, -17, 36
[232] -> Elixir of the Mongoose, -14, 0
[233] -> Sulfuras, Hand of Ragnaros, 0, 80
[234] -> Sulfuras, Hand of Ragnaros, -1, 80
[235] -> Backstage passes to a TAFKAL80ETC concert, -4, 0
[236] -> Backstage passes to a TAFKAL80ETC concert, -9, 0
[237] -> Backstage passes to a TAFKAL80ETC concert, -14, 0
[238] -> Conjured Mana Cake, -16, 0
[239] -> 
[240] -> -------- day 20 --------
[241] -> name, sellIn, quality
[242] -> +5 Dexterity Vest, -10, 0
[243] -> Aged Brie, -18, 38
[244] -> Elixir of the Mongoose, -15, 0
[245] -> Sulfuras, Hand of Ragnaros, 0, 80
[246] -> Sulfuras, Hand of Ragnaros, -1, 80
[247] -> Backstage passes to a TAFKAL80ETC concert, -5, 0
[248] -> Backstage passes to a TAFKAL80ETC concert, -10, 0
[249] -> Backstage passes to a TAFKAL80ETC concert, -15, 0
[250] -> Conjured Mana Cake, -17, 0
[251] -> 
[252] -> -------- day 21 --------
[253] -> name, sellIn, quality
[254] -> +5 Dexterity Vest, -11, 0
[255] -> Aged Brie, -19, 40
[256] -> Elixir of the Mongoose, -16, 0
[257] -> Sulfuras, Hand of Ragnaros, 0, 80
[258] -> Sulfuras, Hand of Ragnaros, -1, 80
[259] -> Backstage passes to a TAFKAL80ETC concert, -6, 0
[260] -> Backstage passes to a TAFKAL80ETC concert, -11, 0
[261] -> Backstage passes to a TAFKAL80ETC concert, -16, 0
[262] -> Conjured Mana Cake, -18, 0
[263] -> 
[264] -> -------- day 22 --------
[265] -> name, sellIn, quality
[266] -> +5 Dexterity Vest, -12, 0
[267] -> Aged Brie, -20, 42
[268] -> Elixir of the Mongoose, -17, 0
[269] -> Sulfuras, Hand of Ragnaros, 0, 80
[270] -> Sulfuras, Hand of Ragnaros, -1, 80
[271] -> Backstage passes to a TAFKAL80ETC concert, -7, 0
[272] -> Backstage passes to a TAFKAL80ETC concert, -12, 0
[273] -> Backstage passes to a TAFKAL80ETC concert, -17, 0
[274] -> Conjured Mana Cake, -19, 0
[275] -> 
[276] -> -------- day 23 --------
[277] -> name, sellIn, quality
[278] -> +5 Dexterity Vest, -13, 0
[279] -> Aged Brie, -21, 44
[280] -> Elixir of the Mongoose, -18, 0
[281] -> Sulfuras, Hand of Ragnaros, 0, 80
[282] -> Sulfuras, Hand of Ragnaros, -1, 80
[283] -> Backstage passes to a TAFKAL80ETC concert, -8, 0
[284] -> Backstage passes to a TAFKAL80ETC concert, -13, 0
[285] -> Backstage passes to a TAFKAL80ETC concert, -18, 0
[286] -> Conjured Mana Cake, -20, 0
[287] -> 
[288] -> -------- day 24 --------
[289] -> name, sellIn, quality
[290] -> +5 Dexterity Vest, -14, 0
[291] -> Aged Brie, -22, 46
[292] -> Elixir of the Mongoose, -19, 0
[293] -> Sulfuras, Hand of Ragnaros, 0, 80
[294] -> Sulfuras, Hand of Ragnaros, -1, 80
[295] -> Backstage passes to a TAFKAL80ETC concert, -9, 0
[296] -> Backstage passes to a TAFKAL80ETC concert, -14, 0
[297] -> Backstage passes to a TAFKAL80ETC concert, -19, 0
[298] -> Conjured Mana Cake, -21, 0
[299] -> 
[300] -> -------- day 25 --------
[301] -> name, sellIn, quality
[302] -> +5 Dexterity Vest, -15, 0
[303] -> Aged Brie, -23, 48
[304] -> Elixir of the Mongoose, -20, 0
[305] -> Sulfuras, Hand of Ragnaros, 0, 80
[306] -> Sulfuras, Hand of Ragnaros, -1, 80
[307] -> Backstage passes to a TAFKAL80ETC concert, -10, 0
[308] -> Backstage passes to a TAFKAL80ETC concert, -15, 0
[309] -> Backstage passes to a TAFKAL80ETC concert, -20, 0
[310] -> Conjured Mana Cake, -22, 0
[311] -> 
[312] -> -------- day 26 --------
[313] -> name, sellIn, quality
[314] -> +5 Dexterity Vest, -16, 0
[315] -> Aged Brie, -24, 50
[316] -> Elixir of the Mongoose, -21, 0
[317] -> Sulfuras, Hand of Ragnaros, 0, 80
[318] -> Sulfuras, Hand of Ragnaros, -1, 80
[319] -> Backstage passes to a TAFKAL80ETC concert, -11, 0
[320] -> Backstage passes to a TAFKAL80ETC concert, -16, 0
[321] -> Backstage passes to a TAFKAL80ETC concert, -21, 0
[322] -> Conjured Mana Cake, -23, 0
[323] -> 
[324] -> -------- day 27 --------
[325] -> name, sellIn, quality
[326] -> +5 Dexterity Vest, -17, 0
[327] -> Aged Brie, -25, 50
[328] -> Elixir of the Mongoose, -22, 0
[329] -> Sulfuras, Hand of Ragnaros, 0, 80
[330] -> Sulfuras, Hand of Ragnaros, -1, 80
[331] -> Backstage passes to a TAFKAL80ETC concert, -12, 0
[332] -> Backstage passes to a TAFKAL80ETC concert, -17, 0
[333] -> Backstage passes to a TAFKAL80ETC concert, -22, 0
[334] -> Conjured Mana Cake, -24, 0
[335] -> 
[336] -> -------- day 28 --------
[337] -> name, sellIn, quality
[338] -> +5 Dexterity Vest, -18, 0
[339] -> Aged Brie, -26, 50
[340] -> Elixir of the Mongoose, -23, 0
[341] -> Sulfuras, Hand of Ragnaros, 0, 80
[342] -> Sulfuras, Hand of Ragnaros, -1, 80
[343] -> Backstage passes to a TAFKAL80ETC concert, -13, 0
[344] -> Backstage passes to a TAFKAL80ETC concert, -18, 0
[345] -> Backstage passes to a TAFKAL80ETC concert, -23, 0
[346] -> Conjured Mana Cake, -25, 0
[347] -> 
[348] -> -------- day 29 --------
[349] -> name, sellIn, quality
[350] -> +5 Dexterity Vest, -19, 0
[351] -> Aged Brie, -27, 50
[352] -> Elixir of the Mongoose, -24, 0
[353] -> Sulfuras, Hand of Ragnaros, 0, 80
[354] -> Sulfuras, Hand of Ragnaros, -1, 80
[355] -> Backstage passes to a TAFKAL80ETC concert, -14, 0
[356] -> Backstage passes to a TAFKAL80ETC concert, -19, 0
[357] -> Backstage passes to a TAFKAL80ETC concert, -24, 0
[358] -> Conjured Mana Cake, -26, 0
[359] -> 
